// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - LivingEdgeLab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package protocol.driver_service;

import "common.proto";

/*
 * Used for low-level control of a vehicle
 */
service Driver {
  // Connects to the vehicle
  rpc Connect (ConnectRequest) returns (ConnectResponse) {}
  // Checks whether the vehicle is successfully connected
  rpc IsConnected (IsConnectedRequest) returns (IsConnectedResponse) {}
  // Disconnects from the vehicle
  rpc Disconnect (DisconnectRequest) returns (DisconnectResponse) {}
  // Order the vehicle to arm
  rpc Arm (ArmRequest) returns (ArmResponse) {}
  // Order the vehicle to spin up
  rpc SpinUp (SpinUpRequest) returns (stream SpinUpResponse) {}
  // Spin down the vehicle at its current position
  rpc SpinDown (SpinDownRequest) returns (stream SpinDownResponse) {}
  // Order the vehicle to hover
  rpc Hold (HoldRequest) returns (stream HoldResponse) {}
  // Emergency shutdown of the vehicle motors
  rpc Kill (KillRequest) returns (stream KillResponse) {}
  // Changes the home destination for the vehicle
  rpc SetHome (SetHomeRequest) returns (SetHomeResponse) {}
  // Return to the vehicle home destination
  rpc ReturnToHome (ReturnToHomeRequest) 
    returns (stream ReturnToHomeResponse) {}
  // Transit the vehicle to a target global position, expressed
  // in global coordinates
  rpc SetGlobalPosition (SetGlobalPositionRequest)
    returns (stream SetGlobalPositionResponse) {}
  // Transit the vehicle to a target position relative to
  // the global ENU (East, North, Up) frame of reference, 
  // in meters
  rpc SetRelativePositionENU (SetRelativePositionENURequest)
    returns (stream SetRelativePositionENUResponse) {}
  // Transit the vehicle to a target position relative to
  // the vehicle frame of reference, in meters
  rpc SetRelativePositionBody (SetRelativePositionBodyRequest) 
    returns (stream SetRelativePositionBodyResponse) {}
  // Transit the vehicle at a target velocity in
  // the global ENU (East, North, Up) frame of reference,
  // in meters per second
  rpc SetVelocityENU (SetVelocityENURequest)
    returns (stream SetVelocityENUResponse) {}
  // Transit the vehicle at a target velocity in
  // the vehicle frame of reference, in meters per second
  rpc SetVelocityBody (SetVelocityBodyRequest)
    returns (stream SetVelocityBodyResponse) {}
  // Turns the vehicle to face a heading or global position
  rpc SetHeading (SetHeadingRequest)
    returns (stream SetHeadingResponse) {}
  // Set the pose of the target gimbal in the global
  // ENU (East, North, Up) frame of reference
  rpc SetGimbalPoseENU (SetGimbalPoseENURequest)
    returns (stream SetGimbalPoseENUResponse) {}
  // Set the pose of the target gimbal in the vehicle
  // frame of reference
  rpc SetGimbalPoseBody (SetGimbalPoseBodyRequest)
    returns (stream SetGimbalPoseBodyResponse) {}
  // Set the vehicle video stream parameters
  rpc ConfigureImagingSensorStream (ConfigureImagingSensorStreamRequest)
    returns (ConfigureImagingSensorStreamResponse) {}
  // Set the vehicle telemetry stream parameters
  rpc ConfigureTelemetryStream (ConfigureTelemetryStreamRequest)
    returns (ConfigureTelemetryStreamResponse) {}   
}

message ConnectRequest {
  protocol.common.Request request = 1;
}

message ConnectResponse {
  protocol.common.Response response = 1;
}

message IsConnectedRequest {
  protocol.common.Request request = 1;
}

message IsConnectedResponse {
  protocol.common.Response response = 1;
  bool is_connected = 2;
}

message DisconnectRequest {
  protocol.common.Request request = 1;
}

message DisconnectResponse {
  protocol.common.Response response = 1;
}

message ArmRequest {
  protocol.common.Request request = 1;
}

message ArmResponse {
  protocol.common.Response response = 1;
}

message SpinUpRequest {
  protocol.common.Request request = 1;
  // Take off height in relative altitude [meters]
  int32 take_off_altitude = 2;
}

message SpinUpResponse {
  protocol.common.Response response = 1;
}

message SpinDownRequest {
  protocol.common.Request request = 1;
}

message SpinDownResponse {
  protocol.common.Response response = 1;
}

message HoldRequest {
  protocol.common.Request request = 1;
}

message HoldResponse {
  protocol.common.Response response = 1;
}

message KillRequest {
  protocol.common.Request request = 1;
}

message KillResponse {
  protocol.common.Response response = 1;
}

message SetHomeRequest {
  protocol.common.Request request = 1;
  protocol.common.Location location = 2; // New home location
}

message SetHomeResponse {
  protocol.common.Response response = 1;
}

message ReturnToHomeRequest {
  protocol.common.Request request = 1;
}

message ReturnToHomeResponse {
  protocol.common.Response response = 1;
}

enum AltitudeMode {
  ABSOLUTE = 0; // Meters above Mean Sea Level
  RELATIVE = 1; // Meters above takeoff location
}

enum HeadingMode {
  TO_TARGET = 0; // Orient towards the target location during transit
  HEADING_START = 1; // Orient towards the given heading, then transit
}

message SetGlobalPositionRequest {
  protocol.common.Request request = 1;
  protocol.common.Location location = 2; // Target location
  // Determines whether the drone will consider altitude as meters above MSL
  // (Mean Sea Level), or relative to its takeoff location.
  AltitudeMode altitude_mode = 4;
  // Determines how the drone will orient during transit 
  HeadingMode heading_mode = 5;
  // Maximum velocity during transit, north_vel determines horizontal
  // velocity, up_vel determines vertical velocity, and angular_vel
  // determines angular velocity
  protocol.common.VelocityENU max_velocity = 6;
}

message SetGlobalPositionResponse {
  protocol.common.Response response = 1;
}

message SetRelativePositionENURequest {
  protocol.common.Request request = 1;
  // Target position in the ENU coordinate frame
  protocol.common.PositionENU position = 2;
  // Maximum velocity during transit, north_vel determines horizontal
  // velocity, up_vel determines vertical velocity, and angular_vel
  // determines angular velocity
  protocol.common.VelocityENU max_velocity = 3;
}

message SetRelativePositionENUResponse {
  protocol.common.Response response = 1;
}

message SetRelativePositionBodyRequest {
  protocol.common.Request request = 1;
  // Target position in the body coordinate frame
  protocol.common.PositionBody position = 2;
  // Maximum velocity during transit, forward_vel determines horizontal
  // velocity, up_vel determines vertical velocity, and angular_vel
  // determines angular velocity
  protocol.common.VelocityBody max_velocity = 3;
}

message SetRelativePositionBodyResponse {
  protocol.common.Response response = 1;
}

message SetVelocityENURequest {
  protocol.common.Request request = 1;
  // Velocity in the ENU coordinate frame
  protocol.common.VelocityENU velocity = 2;
}

message SetVelocityENUResponse {
  protocol.common.Response response = 1;
}

message SetVelocityBodyRequest {
  protocol.common.Request request = 1;
  // Velocity in the body coordinate frame
  protocol.common.VelocityBody velocity = 2;
}

message SetVelocityBodyResponse {
  protocol.common.Response response = 1;
}

message SetHeadingRequest {
  protocol.common.Request request = 1;
  protocol.common.Location location = 2;
  // Determines whether to use provided heading (HEADING_START)
  // or to look at target GPS (TO_TARGET)
  HeadingMode heading_mode = 5;
}

message SetHeadingResponse {
  protocol.common.Response response = 1;
}

enum GimbalMode {
  // Provided pose is absolute
  ABSOLUTE_ANGLE = 0;
  // Provided pose is relative to current pose
  RELATIVE_ANGLE = 1;
  // Provided values are angular velocities
  ANGULAR_VELOCITY = 2;
}

message SetGimbalPoseBodyRequest {
  protocol.common.Request request = 1;
  uint32 gimbal_id = 2; // ID of the target gimbal
  protocol.common.Pose pose = 3; // Target pose in the body coordinate frame
  GimbalMode mode = 4; // Specifies how to interpret the target pose
}

message SetGimbalPoseBodyResponse {
  protocol.common.Response response = 1;
}

message SetGimbalPoseENURequest {
  protocol.common.Request request = 1;
  uint32 gimbal_id = 2; // ID of the target gimbal
  protocol.common.Pose pose = 3; // Target pose in the ENU coordinate frame
  GimbalMode mode = 4; // Specifies how to interpret the target pose
}

message SetGimbalPoseENUResponse {
  protocol.common.Response response = 1;
}

message ImagingSensorConfiguration {
  uint32 id = 1; // Target imaging sensor ID
  bool set_primary = 2; // Set this sensor as the primary stream
  uint32 set_fps = 3; // Target FPS for stream
} 

message ConfigureImagingSensorStreamRequest {
  protocol.common.Request request = 1;
  // Channel to send frames on
  string channel = 2;
  // List of configurations to be updated
  repeated ImagingSensorConfiguration configurations = 3;
}

message ConfigureImagingSensorStreamResponse {
  protocol.common.Response response = 1;
}

message ConfigureTelemetryStreamRequest {
  protocol.common.Request request = 1;
  // Channel to send telemetry on
  string channel = 2;
  // Target frequency of telemetry generation, in Hz
  uint32 frequency = 3;
}

message ConfigureTelemetryStreamResponse {
  protocol.common.Response response = 1;
}
